### PR TITLE
Use `CcrRepository` to init follower index (#35719)

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit modifies the put follow index action to use a
CcrRepository when creating a follower index. It routes
the logic through the snapshot/restore process. A
wait_for_active_shards parameter can be used to configure
how long to wait before returning the response.

### CODE REVIEW

**Comment by tbrooks8** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by tbrooks8** [comment]
@ywelsch I added a pre-6.7 put follow mode. The code is copied from the prior `TransportPutFollowAction`.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java [hunk 1]
```diff
@@ -132,39 +142,49 @@ private void createFollowerIndex(
...
-            .indices(request.getLeaderIndex()).indicesOptions(request.indicesOptions()).renamePattern("^(.*)$")
-            .renameReplacement(request.getFollowRequest().getFollowerIndex()).masterNodeTimeout(request.masterNodeTimeout())
-            .indexSettings(settingsBuilder);
+        boolean pre67CompatibilityMode = localClusterMinNodeVersion.before(Version.V_6_7_0)
+            || metaDataTuple.v1().getNodes().getMinNodeVersion().before(Version.V_6_7_0);
 
-        final Client clientWithHeaders = CcrLicenseChecker.wrapClient(this.client, threadPool.getThreadContext().getHeaders());
-        threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(new AbstractRunnable() {
+        if (pre67CompatibilityMode) {
+            logger.warn("Pre-6.7 nodes present in local/remote cluster. Cannot bootstrap from remote. Creating empty follower index [{}] " +
```
Comment (ywelsch) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java]: why log a warning? Isn't this just a normal thing? Log at debug level here.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrLicenseChecker.java [hunk 2]
```diff
@@ -134,7 +135,7 @@ public void checkRemoteClusterLicenseAndFetchLeaderIndexMetadataAndHistoryUUIDs(
                     hasPrivilegesToFollowIndices(remoteClient, new String[] {leaderIndex}, e -> {
                         if (e == null) {
                             fetchLeaderHistoryUUIDs(remoteClient, leaderIndexMetaData, onFailure, historyUUIDs ->
-                                    consumer.accept(historyUUIDs, leaderIndexMetaData));
+                                    consumer.accept(historyUUIDs, new Tuple<>(remoteClusterState, leaderIndexMetaData)));
```
Comment (ywelsch) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrLicenseChecker.java]: Don't you need to set `.nodes(true)` on the cluster state request in order to get the minNodeVersion? Why is this not failing any tests?

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've only reviewed eea717e (the extra changes for 6.7) and left 2 comments.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrLicenseChecker.java (hunk 2 above)
Comment (tbrooks8) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrLicenseChecker.java]: `nodes` defaults to true. But I will set it explicitly.

Comment (tbrooks8) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrLicenseChecker.java]: Oh well I see that we set `clear()` which should reset it. I'll investigate why the test did not fail.

**Comment by tbrooks8** [comment]
@elasticmachine run elasticsearch-ci/2

**Review by martijnvg — APPROVED** [review]
LGTM

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java (hunk 1 above)
Comment (ywelsch) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java]: I think the same applies to any feature we introduce in a new version. If you are in a mixed version cluster, the feature will not work. The special case here is that a remote cluster might be on an older version, not supporting that feature. I think this is a natural extension of said logic. Just as for CCS, if you search over multiple cluster with a feature introduced in a newer version, that will not work in the presence of older clusters.

**Review by ywelsch — APPROVED** [review]
LGTM

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/Pre67PutFollow.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.xpack.ccr.Ccr;
+import org.elasticsearch.xpack.ccr.CcrSettings;
+import org.elasticsearch.xpack.core.ccr.action.PutFollowAction;
+import org.elasticsearch.xpack.core.ccr.action.ResumeFollowAction;
+
+import java.util.HashMap;
+import java.util.Map;
+
+final class Pre67PutFollow {
```
[the remark this answers is not in the document]
Comment (tbrooks8) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/Pre67PutFollow.java]: Yes. Not always but sometimes.

**Pull request merged**